### PR TITLE
audit(identity): record userId + apiKeyId as separate fields (refs #97)

### PR DESCRIPTION
**Pull request by alex-budanov**
Refs [#97](https://github.com/apex-bridge/bugspotter/issues/97) — audit-identity masking when JWT and full-scope API key are present in the same request. **Partial fix; #97 stays open for the dual-header user-attribution gap (see "Honest scope" below).**

## The gap

Auth middleware short-circuits on `x-api-key` and never sets `request.authUser` ([auth/middleware.ts:71](packages/backend/src/api/middleware/auth/middleware.ts#L71)). Downstream loggers wrote `userId: authUser?.id || 'api-key'` — the literal `'api-key'` masked the actual JWT user's identity in audit logs. A user with access to an org's full-scope key could deliberately attach it to JWT-authenticated requests to obscure attribution.

## Why Option B

[#97](https://github.com/apex-bridge/bugspotter/issues/97) outlined three fixes; this ships **Option B** (record both identities). The reasoning is in the issue and in the PR conversation, but the short version: Options A and C both touch the auth-decision predicate (`apiKey && !authUser`) used in `checkProjectAccess` to take the api-key branch — small middleware change, large blast radius if any callsite isn't audited carefully. B is purely additive: existing audit consumers reading `userId` keep working; `apiKeyId` is new.

## Honest scope — what this PR closes vs. what stays open

This PR closes part of #97 and intentionally leaves part of it open:

| Case | Before | After |
|---|---|---|
| api-key only | `user_id: 'api-key'` (literal masquerading as a userId) | `user_id: null`, `details.api_key_id: <key id>` |
| JWT only | `user_id: <jwt user>` | `user_id: <jwt user>`, `details.api_key_id: null` |
| **Dual-header (JWT + api-key)** | `user_id: 'api-key'` (human attribution masked, machine attribution masked) | `user_id: null`, `details.api_key_id: <key id>` — **machine attribution captured, but JWT user's id is still null** |

The dual-header case still leaves the human actor unattributed, because the auth middleware never populates `request.authUser` when an api-key is present. PR-105 fixes the worst of the masking (the `'api-key'` literal) and adds first-class machine attribution; the underlying auth-flow change (run JWT parsing alongside api-key validation for *identity* purposes — separate from the precedence decision) is out of scope and is what keeps #97 open.

A follow-up that closes the dual-header gap would (a) leave header precedence unchanged for authz, but (b) populate a `request.jwtUserIdentity` (or similar) in the api-key path so audit consumers can record both. That's a deeper middleware change with its own design trade-offs (claim verification cost on every api-key request, etc.) — worth a focused PR.

## What changes

| Change | Where |
|---|---|
| 13 handler logger sites — `userId: authUser?.id \|\| 'api-key'` → `userId/apiKeyId` pair | `src/api/routes/integrations.ts`, `src/api/routes/integration-rules.ts` |
| Global audit middleware writes `api_key_id` in `details` JSONB | `src/api/middleware/audit.ts` |
| `createAuditLog` helper requires `apiKeyId: string \| null` (no silent omission) and always emits `api_key_id` for uniform JSONB shape | `src/api/middleware/audit.ts` |
| ESLint rule bans `Property:matches([key.name='userId'], [key.value='userId']) Literal[value='api-key']` (catches both quoted and unquoted keys) | root `eslint.config.js` |
| Runtime regression coverage — three tests asserting audit-row shape on api-key-only / JWT-only / dual-header writes | `tests/api/audit-middleware.test.ts` |
| Doc + comment updates | `packages/backend/docs/auth.md`, `tests/integration/full-scope-api-key.test.ts` precedence-test comment, `audit.ts` middleware comment |

## Tracked separately

- [#97](https://github.com/apex-bridge/bugspotter/issues/97) stays open with narrowed scope: dual-header user attribution (the final gap above).
- [#104](https://github.com/apex-bridge/bugspotter/issues/104) — add a first-class `api_key_id` column to `audit_logs` for indexed api-key-attributed queries. Today the data lives in `details` JSONB (no schema change needed); column-level indexing follows in a focused migration PR.

## Test plan

- [x] Typecheck clean
- [x] Backend API test suite: 1905/1905 passing locally (was 1902 — added 3 audit-identity assertion tests)
- [x] ESLint rule verified — fires on all four forms (`userId: x \|\| 'api-key'`, `userId: 'api-key'`, quoted-key variants), doesn't false-positive on the integration-plugin auth-type union (`type: 'api-key'`) or sensitive-header lists
- [x] Runtime regression coverage — `tests/api/audit-middleware.test.ts > API Key Identity Recording` asserts on the produced audit row for all three identity cases, so a refactor that drops `api_key_id` from `auditData.details` (or re-introduces a masked literal) would surface as a red test, not just a green build with bad data

## Notes

- **Audit consumer impact**: any external system that reads logs and expects `userId` to always be a non-null string needs to handle `userId: null` with `apiKeyId` populated. Internal log queries should adapt; SIEM/BI integrations may need a one-line update
- **No middleware change**: header precedence is unchanged (api-key still wins for authz). Only the audit fields differ. Documented in `docs/auth.md` §audit-identity

Generated with Claude Code

## Summary by CodeRabbit

* **Bug Fixes**
  * Audit logs now record user identity and API key identity as separate fields instead of using a literal fallback, improving attribution.

* **Documentation**
  * Auth docs updated to reflect the partial fix and clarify remaining dual-header behavior where user attribution can still be null.

* **Linting**
  * New ESLint rule prevents using the literal 'api-key' as a user identifier.

* **Tests**
  * Added tests verifying API-key and JWT identity recording; updated test comments.

### CODE REVIEW

File: packages/backend/src/api/middleware/audit.ts [hunk 1]
```diff
@@ -183,9 +183,33 @@ export function createAuditMiddleware(db: DatabaseClient) {
...
+    //     null. Attribution captured.
+    //   - DUAL-HEADER (JWT + api-key): `user_id` STILL null,
+    //     `details.api_key_id` populated. The auth middleware
+    //     short-circuits on api-key (auth/middleware.ts:71) and never
+    //     populates `request.authUser`, so the JWT user's id is
+    //     unreachable from this hook. The audit trail therefore
+    //     captures the machine identity but loses the human actor —
+    //     the same gap GH-97 named, only partially closed by the
+    //     handler-level loggers (which compute `userId` from the route
+    //     context, not from `request.authUser`). Closing it here
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts]: Fair catch — fixed in e3e826b. The handler-level loggers in integrations.ts/integration-rules.ts use `request.authUser?.id ?? null` exactly like this middleware, so they share the dual-header gap rather than partially closing it. Comment now reflects that.

File: eslint.config.js [hunk 2]
```diff
@@ -44,6 +44,19 @@ export default tseslint.config(
...
+      // logs or audit_log row writes. Record `userId` and `apiKeyId` as
+      // separate fields so dual-header (JWT + api-key) requests attribute
+      // correctly. See packages/backend/docs/auth.md §audit-identity.
+      'no-restricted-syntax': [
+        'error',
+        {
+          selector: "Property[key.name='userId'] Literal[value='api-key']",
+          message:
+            "Don't use the 'api-key' literal as a userId. Record `userId: authUser?.id ?? null` and `apiKeyId: apiKey?.id ?? null` as separate fields. See packages/backend/docs/auth.md §audit-identity.",
+        },
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply eslint.config.js]: Good catch — fixed in 94b9168. Selector is now `Property:matches([key.name='userId'], [key.value='userId']) Literal[value='api-key']` so it catches all four forms: `{ userId: 'api-key' }`, `{ 'userId': 'api-key' }`, `{ userId: x || 'api-key' }`, `{ 'userId': x || 'api-key' }`. The previous `key.name`-only selector would have let quoted keys slip past.

File: packages/backend/src/api/middleware/audit.ts [hunk 3]
```diff
@@ -183,9 +183,15 @@ export function createAuditMiddleware(db: DatabaseClient) {
       return;
     }
 
-    // Get user info from auth context
+    // Get user info from auth context. Both `authUser` and `apiKey`
+    // are recorded so that audit trails attribute api-key-only and
+    // dual-header (JWT + api-key) requests correctly. The
+    // `audit_logs.user_id` column stays null for api-key-only requests
+    // (see GH-97 for the deferred schema-level api_key_id column);
+    // until then, the api-key id rides along in the `details` JSONB.
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts]: You're right — the original wording oversold what this PR closes. Fixed in 94b9168 (and follow-up e3e826b). The comment now enumerates the three cases honestly:

- api-key-only: `user_id` null, `details.api_key_id` populated
- JWT-only: `user_id` populated, `details.api_key_id` null
- DUAL-HEADER (JWT + api-key): `user_id` STILL null because the auth middleware short-circuits on api-key (auth/middleware.ts:71) and never populates `request.authUser`

PR-105 is Option B — it gives api-key requests a first-class attribution field and bans the `'api-key'` literal as a userId. Closing the dual-header gap requires a deeper auth-middleware change (run JWT parsing alongside api-key validation for identity) and is out of scope here.

File: packages/backend/src/api/middleware/audit.ts [hunk 4]
```diff
@@ -252,7 +290,10 @@ export async function createAuditLog(
       action: data.action,
       resource: data.resource,
       resource_id: data.resourceId || null,
-      details: data.details || null,
+      details:
+        data.apiKeyId !== undefined
+          ? { ...(data.details ?? {}), api_key_id: data.apiKeyId }
+          : (data.details ?? null),
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts]: Good catch — fixed in c5a63d1. Two changes addressing the same root issue:

1. `apiKeyId` is now required (`string | null`) so callers must explicitly pass `null` for human-only actions. Verified zero current callers exist (grep confirmed only `createAuditLogger` from the api-key service is referenced; `createAuditLog` from audit.ts has no direct consumers), so it's a defensive change rather than a breaking one — but a future caller forgetting the field would have produced rows with no `api_key_id` at all.
2. The helper now always emits `api_key_id` in details (matching the global middleware), giving the GH-104 column-level migration a uniform JSONB shape to backfill from.

File: packages/backend/tests/integration/full-scope-api-key.test.ts [hunk 5]
```diff
@@ -711,19 +711,19 @@ describe('Full-Scope API Key Integration Tests', () => {
...
+      // on this dual-header path, downstream audit logs record
+      // `userId: null, apiKeyId: <key id>`. Combining JWT + api-key
+      // therefore no longer masks user attribution by collapsing it to
+      // the machine identity; the api-key id is recorded explicitly
+      // and the user-side null is honest about what the middleware
+      // saw. If a future PR flips the middleware to prefer JWT (or
+      // populates `authUser` even when an api-key is present), audit
+      // logs would start recording both ids, which is the correct
+      // outcome under either precedence model.
       expect(response.statusCode).toBe(200);
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/tests/integration/full-scope-api-key.test.ts:727]: Solid point about the assertion gap — done in c5a63d1, but with two clarifications:

1. **The suggested fix wouldn't have worked on this specific test.** `AUDIT_METHODS` in audit.ts:22 is `{POST, PUT, PATCH, DELETE}` — GETs are deliberately not audited. The dual-header test here is a GET on `/api/v1/screenshots/{id}`, so no audit row is ever produced and `findLatestForRequest` (which also doesn't exist on the audit-log repo) wouldn't find anything to assert on.

2. **Added the coverage in the right place instead.** Three new tests in `tests/api/audit-middleware.test.ts > API Key Identity Recording` use `PATCH /api/v1/reports/:id` (auditable + accepts api-key auth) and assert on the actual audit row for all three identity cases:
   - api-key only → `user_id: null`, `details.api_key_id: <key id>`
   - dual-header (JWT + api-key) → same shape as api-key-only (the GH-97 attack-shape guard)
   - JWT only → `user_id: <user>`, `details.api_key_id: null`

A regression dropping `api_key_id` from `auditData.details` or re-introducing the masked `'api-key'` literal would now turn one of these red instead of leaving green CI with bad data. The original test you flagged (precedence test in `full-scope-api-key.test.ts`) is still doing its job — verifying the auth-middleware short-circuit — and the audit-row claim now has its own dedicated coverage.

File: packages/backend/src/api/middleware/audit.ts [hunk 6]
```diff
@@ -183,9 +183,34 @@ export function createAuditMiddleware(db: DatabaseClient) {
...
+    //     unreachable from this hook. The audit trail therefore
+    //     captures the machine identity but loses the human actor.
+    //     Handler-level loggers (e.g. routes/integrations.ts,
+    //     routes/integration-rules.ts) read the same
+    //     `request.authUser?.id ?? null`, so they share this gap —
+    //     they do NOT partially close it. Closing it here requires
+    //     a deeper auth-middleware change that runs JWT parsing
+    //     alongside api-key validation for identity purposes; out
+    //     of scope for #97's Option B fix and warrants its own
+    //     design pass.
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts:206]: You're right — fully agreed and acted on:

- **PR title** updated from `(closes #97)` → `(refs #97)` so this doesn't auto-close the issue on merge.
- **PR description** rewritten with an honest scope table that enumerates all three cases (api-key-only / JWT-only / dual-header). Dual-header still records `user_id: null` because the auth middleware short-circuits on api-key and never populates `request.authUser`. PR-105 fixes the literal-masking and adds first-class machine attribution; the human-attribution gap on dual-header is what keeps #97 open.
- **The follow-up direction** is sketched in the new "Honest scope" section: leave header precedence unchanged for authz, but populate something like `request.jwtUserIdentity` in the api-key path so audit consumers can record both. That's a deeper middleware change with its own design trade-offs (claim verification cost on every api-key request) — its own PR.

So #97 stays open with a narrowed scope, and PR-105 is now framed as a partial fix rather than a closure.

File: packages/backend/src/api/middleware/audit.ts [hunk 7]
```diff
@@ -238,6 +269,13 @@ export async function createAuditLog(
   db: DatabaseClient,
   data: {
     userId?: string;
+    /**
+     * API-key id when the action was driven by a machine credential
+     * (or by a user with an api-key in the same request — see GH-97
+     * audit-identity gap). Both fields are recorded so dual-header
+     * requests attribute correctly.
+     */
+    apiKeyId?: string;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts]: Agreed and applied in c5a63d1. `apiKeyId` is now `string | null` (required) instead of optional. The TypeScript signature now forces every caller to make an explicit choice, so omission is a compile error.

Verified the blast radius before tightening: grepped `createAuditLog` across the backend src and the only matches outside audit.ts itself are calls to `this.createAuditLog` (a private method on `RetentionService`, unrelated) and `createAuditLogger` (a different export from `services/api-key/audit-logger.ts`). Zero current callers of the audit.ts helper, so this is a purely defensive change — but it permanently closes the silent-omission shape for any future caller.

File: packages/backend/docs/auth.md [hunk 8]
```diff
@@ -160,7 +160,7 @@ design — but they're known design choices that may change. New work
 adjacent to these should consult the linked issues before assuming
 permanence.
 
-- **[#97](https://github.com/apex-bridge/bugspotter/issues/97) — Audit-identity masking**: JWT + full-scope API key obscures user attribution in logs because `request.authUser` is never set when an API key is present. Three fix options proposed (populate `authUser` anyway, log both identities, or reject dual-header). Decision needed.
+- ~~**[#97](https://github.com/apex-bridge/bugspotter/issues/97) — Audit-identity masking**~~: closed. Handler-level audit logs now record `userId: authUser?.id ?? null` and `apiKeyId: apiKey?.id ?? null` as separate fields (Option B from the issue). The global audit middleware writes the same pair — `user_id` to its dedicated column, `api_key_id` to the `details` JSONB. A schema-level `audit_logs.api_key_id` column is deferred (tracked separately) so api-key-attributed queries can index without scanning JSONB. ESLint rule `no-restricted-syntax` (root config) prevents regression on the `'api-key'` literal pattern.
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/docs/auth.md]: Good catch on the inconsistency — fixed in 54763a9. auth.md §6 now matches the PR description and the audit.ts middleware comment: "partial fix; #97 stays open with narrowed scope," with the dual-header user-attribution gap explicitly called out.

Used a slightly different shape than your suggested diff so it parallels the description's narrative instead of restating the change-list (the doc's §6 "Open questions" is a design-intent log; what's *not* yet done belongs there at least as much as what landed).

File: packages/backend/src/api/middleware/audit.ts [hunk 9]
```diff
@@ -252,7 +294,10 @@ export async function createAuditLog(
       action: data.action,
       resource: data.resource,
       resource_id: data.resourceId || null,
-      details: data.details || null,
+      // Always include `api_key_id` in details (matching the global
+      // middleware) so the column-level migration in GH-104 can
+      // backfill from a uniform JSONB shape.
+      details: { ...(data.details ?? {}), api_key_id: data.apiKeyId },
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts:319]: Real concern, partly addressed in 54763a9. Two parts:

**The contract change is intentional, not accidental.** PR-105 deliberately moves `details` to always-non-null because the GH-104 column-level migration needs a uniform JSONB shape to backfill from. The global middleware already always wrote `details: { body, query, params, api_key_id }`; the helper now matches that shape. So `details IS NULL` semantics for "this audit row has no extra metadata" are gone for any row written through the audit subsystem — that's the desired end-state, not a regression.

**Verified the blast radius is zero today.** Grepped `details IS NULL` / `details = null` / `details === null` across `packages/backend/src` (`.ts` and `.sql`) and got no matches. Combined with zero direct callers of `createAuditLog`, no current path is affected.

**What I did, and what I deliberately didn't.** Documented the always-non-null contract in the JSDoc (54763a9), including the implication for SIEM rules / callers — `WHERE details IS NULL` will never match; filter on `details ? 'api_key_id'` or `details->>'api_key_id'` instead. Did **not** add a unit test for createAuditLog directly: with zero callers and a deliberate contract change, a unit test would only assert the contract that the JSDoc now states. If a real caller is added in a future PR, that PR should bring the test along with it.

Happy to add the unit test now if you'd rather have belt-and-suspenders coverage on the helper before any caller exists — flag and I'll push a follow-up.

File: packages/backend/src/api/middleware/audit.ts [hunk 10]
```diff
@@ -231,13 +262,34 @@ export function createAuditMiddleware(db: DatabaseClient) {
...
+ * migration (api_key_id moves out of JSONB) and matches the global
+ * middleware's behaviour. **Implication for callers / SIEM rules**:
+ * `WHERE details IS NULL` will never match rows written by this
+ * helper; filter on `details ? 'api_key_id'` or `details->>'api_key_id'`
+ * instead if you want "no machine attribution" semantics.
  */
 export async function createAuditLog(
   db: DatabaseClient,
   data: {
     userId?: string;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/middleware/audit.ts]: Excellent point — fully agreed and applied in 95dfda0. `userId` is now `string | null` (required), symmetric with `apiKeyId`. The exact same silent-omission rationale applies: a caller writing `{ apiKeyId: null, action, resource }` and omitting `userId` silently dropped human attribution with no compile signal. Re-verified zero current direct callers via grep, so the change is purely defensive — but it permanently closes the asymmetry for any future caller. Thanks for catching that the rationale was load-bearing on both fields, not just one.

File: packages/backend/tests/api/audit-middleware.test.ts [hunk 11]
```diff
@@ -460,4 +461,133 @@ describe('Audit Middleware', () => {
...
+      expect(details.api_key_id).toBe(key.id);
+    });
+
+    it('records api_key_id and user_id null on dual-header (JWT + api-key) request', async () => {
+      // Auth middleware short-circuits on the api-key header
+      // (auth/middleware.ts:71) and never populates request.authUser even
+      // when a JWT is present, so the audit row mirrors the api-key-only
+      // case. This is the GH-97 attack-shape guard: a regression that
+      // "fixes" the precedence by reading the JWT alongside the api-key
+      // would surface here as user_id no longer being null.
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/tests/api/audit-middleware.test.ts]: Right, the framing was misleading and that's a real footgun for whoever lands #97's follow-up. Fixed in 95dfda0. The comment now says explicitly:

> When #97's follow-up lands, this test WILL fail — that's expected. The follow-up is supposed to populate `user_id` from the JWT alongside the api-key for *identity* purposes (without changing authz precedence). At that point, invert the assertion to `expect(log.user_id).toBe(testUserId)` and update the comment. Don't revert the change or delete the assertion — that would re-open the human-attribution gap.

So a developer who hits the failure now sees both the *what* (invert the assertion) and the *why not just delete it* (deleting re-opens the gap). The assertion still pins the current limitation; only the framing changed.

File: packages/backend/tests/api/audit-middleware.test.ts [hunk 12]
```diff
@@ -460,4 +461,133 @@ describe('Audit Middleware', () => {
...
+    async function waitForReportAudit(reportId: string) {
+      for (let i = 0; i < 30; i++) {
+        const log = await findLatestReportAudit(reportId);
+        if (log) {
+          return log;
+        }
+        await new Promise((r) => setTimeout(r, 50));
+      }
+      throw new Error(`audit row for /api/v1/reports/${reportId} not written within 1.5s`);
+    }
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/tests/api/audit-middleware.test.ts:523]: Two-part response in 95dfda0:

**On flake risk:** bumped to 60 × 50ms = 3s. Borderline runners shouldn't trip. Improved the error message so a future flake surfaces *what* failure mode it implies — fire-and-forget audit write swallowed an error in its `.catch`, or a regression dropped the call entirely — rather than the contentless "not written within Ns."

**On false-confidence (the deeper concern):** you're right that the test would pass vacuously if the audit write silently dropped. The poll-and-fail timeout *does* fail the test in that case (so it's not entirely vacuous), but the failure happens 3s late and the error message is what a developer reads to figure out whether it's flake or a real bug. Improving the message addresses the diagnostic gap; the underlying reliability gap (audit middleware uses fire-and-forget on `onResponse` with only a log line on failure — audit.ts:258) is pre-existing and genuinely out of scope for PR-105.

Worth tracking the reliability gap separately — "audit writes need retry / DLQ to ensure attribution survives transient DB errors" is its own design discussion (idempotency, ordering relative to the response, queue choice). Filing it as a follow-up issue rather than letting it ride here.

File: eslint.config.js [hunk 13]
```diff
@@ -44,6 +44,25 @@ export default tseslint.config(
...
+      // The selector uses `:matches()` so it catches both forms of the
+      // `userId` property: unquoted (Identifier — `key.name`) and
+      // quoted (Literal — `key.value`). Without `:matches()` only the
+      // unquoted form is detected and a developer writing
+      // `{ 'userId': 'api-key' }` would slip past.
+      'no-restricted-syntax': [
+        'error',
+        {
+          selector:
+            "Property:matches([key.name='userId'], [key.value='userId']) Literal[value='api-key']",
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply eslint.config.js]: Good catch — fixed in 6f27e49. Selector now wraps both legs in `:matches()` so the four regression shapes are all covered:

| Shape | Node type |
|---|---|
| `{ userId: 'api-key' }` | Literal |
| `{ 'userId': 'api-key' }` | Literal (quoted key) |
| <code>{ userId: &#96;api-key&#96; }</code> | zero-interp TemplateLiteral |
| <code>{ 'userId': &#96;api-key&#96; }</code> | zero-interp TemplateLiteral (quoted key) |

Verified locally with a fixture that includes all four shapes plus a near-miss (`'api-key2'` — correctly not flagged). Same runtime semantics, same risk, now uniformly covered.

While unlikely in practice (no current code uses backticks for static strings here), the selector is the regression-guard contract — covering the AST shape and not just "the form code happens to take today" makes it robust against routine refactors that change quoting style.
